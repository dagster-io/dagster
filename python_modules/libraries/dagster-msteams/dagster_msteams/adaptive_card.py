from typing import Any


class AdaptiveCard:
    """Class to contruct a MS Teams adaptive card for posting Dagster messages."""

    def __init__(self, adaptive_card_version: str = "1.5"):
        """Constructs an adaptive card with the given version.

        Args:
            adaptive_card_version (str): The version of the adaptive card to use. Defaults to "1.5".
        """
        self._body = []
        self._adaptive_card_version = adaptive_card_version

    @property
    def payload(self) -> dict[str, Any]:
        return {
            "type": "message",
            "attachments": [
                {
                    "contentType": "application/vnd.microsoft.card.adaptive",
                    "contentUrl": None,
                    "content": {
                        "type": "AdaptiveCard",
                        "version": self._adaptive_card_version,
                        "body": self._body,
                    },
                }
            ],
        }

    def _build_text_block(self, text: str) -> dict[str, Any]:
        return {
            "type": "TextBlock",
            "text": text,
            "wrap": True,
        }

    def add_attachment(self, text_message: str) -> None:
        """Appends a text message to the adaptive card."""
        self._body.append(self._build_text_block(text_message))
