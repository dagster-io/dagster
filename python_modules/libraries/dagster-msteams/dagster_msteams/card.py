from collections.abc import Mapping


class Card:
    """Class to contruct a MS Teams Hero card for posting dagster messages."""

    def __init__(self):
        self.payload = {"type": "message", "attachments": []}

    @property
    def type(self):
        return self.payload["type"]

    @property
    def attachments(self):
        return self.payload["attachments"]

    def _create_attachment(self, text_message: str) -> Mapping:
        content = {
            "title": "Dagster Pipeline Alert",
            "text": text_message,
        }
        content_type = "application/vnd.microsoft.card.hero"
        return {"contentType": content_type, "content": content}

    def add_attachment(self, text_message: str):
        hero_card_attachment = self._create_attachment(text_message)
        self.payload["attachments"].append(hero_card_attachment)
