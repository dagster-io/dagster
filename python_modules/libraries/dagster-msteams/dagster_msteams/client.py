from collections.abc import Mapping
from typing import Optional

from requests import codes, exceptions, post


class TeamsClient:
    """MS Teams web client responsible for connecting to a channel using the webhook URL
    and posting informaton in the form of cards.
    """

    def __init__(
        self,
        hook_url: str,
        http_proxy: Optional[str] = None,
        https_proxy: Optional[str] = None,
        timeout: Optional[float] = 60,
        verify: Optional[bool] = None,
    ):
        self._hook_url = hook_url
        self._timeout = timeout
        self._verify = verify
        if http_proxy is None and https_proxy is None:
            self._proxy = None
        else:
            self._proxy = {}
            if http_proxy:
                self._proxy["http"] = http_proxy
            if https_proxy:
                self._proxy["https"] = https_proxy
        self._headers = {"Content-Type": "application/json"}

    def post_message(self, payload: Mapping) -> bool:  # pragma: no cover
        response = post(
            self._hook_url,
            json=payload,
            headers=self._headers,
            proxies=self._proxy,
            timeout=self._timeout,
            verify=self._verify,
        )
        if response.status_code == codes["ok"] and response.text == "1":
            return True
        else:
            raise exceptions.RequestException(response.text)
