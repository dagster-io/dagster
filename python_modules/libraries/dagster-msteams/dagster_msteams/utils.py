from typing import NamedTuple, Optional


class MSTeamsHyperlink(NamedTuple):
    text: str
    url: str


def build_message_with_link(
    is_legacy_webhook: bool, text: str, link: Optional[MSTeamsHyperlink]
) -> str:
    if link:
        if is_legacy_webhook:
            return f"{text} <a href='{link.url}'>{link.text}</a>"
        else:
            return f"{text} [{link.text}]({link.url})"

    return text
