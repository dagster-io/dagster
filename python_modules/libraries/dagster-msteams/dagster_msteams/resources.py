from typing import Optional

from dagster import ConfigurableResource, resource
from dagster._core.definitions.resource_definition import dagster_maintained_resource
from pydantic import Field

from dagster_msteams.client import TeamsClient


class MSTeamsResource(ConfigurableResource):
    """This resource is for connecting to Microsoft Teams.

    Provides a `dagster_msteams.TeamsClient` which can be used to
    interface with the MS Teams API.

    By configuring this resource, you can post messages to MS Teams from any Dagster op,
    asset, schedule, or sensor:

    Examples:
        .. code-block:: python

            import os

            from dagster import op, job, Definitions, EnvVar
            from dagster_msteams import Card, MSTeamsResource


            @op
            def teams_op(msteams: MSTeamsResource):
                card = Card()
                card.add_attachment(text_message="Hello There !!")
                msteams.get_client().post_message(payload=card.payload)


            @job
            def teams_job():
                teams_op()

            Definitions(
                jobs=[teams_job],
                resources={
                    "msteams": MSTeamsResource(
                        hook_url=EnvVar("TEAMS_WEBHOOK_URL")
                    )
                }
            )
    """

    hook_url: str = Field(
        description=(
            "To send messages to MS Teams channel, an incoming webhook has to be created. The"
            " incoming webhook url must be given as a part of the resource config to the"
            " MSTeamsResource in Dagster. For more information on how to create an incoming"
            " webhook, see"
            " https://docs.microsoft.com/en-us/microsoftteams/platform/webhooks-and-connectors/how-to/add-incoming-webhook"
        ),
    )
    http_proxy: Optional[str] = Field(default=None, description="HTTP proxy URL")
    https_proxy: Optional[str] = Field(default=None, description="HTTPS proxy URL")
    timeout: float = Field(default=60, description="Timeout for requests to MS Teams")
    verify: bool = Field(
        default=True, description="Whether to verify SSL certificates, defaults to True"
    )

    @classmethod
    def _is_dagster_maintained(cls) -> bool:
        return True

    def get_client(self) -> TeamsClient:
        return TeamsClient(
            hook_url=self.hook_url,
            http_proxy=self.http_proxy,
            https_proxy=self.https_proxy,
            timeout=self.timeout,
            verify=self.verify,
        )


@dagster_maintained_resource
@resource(
    config_schema=MSTeamsResource.to_config_schema(),
    description="This resource is for connecting to MS Teams",
)
def msteams_resource(context) -> TeamsClient:
    """This resource is for connecting to Microsoft Teams.

    The resource object is a `dagster_msteams.TeamsClient`.

    By configuring this resource, you can post messages to MS Teams from any Dagster solid:

    Examples:
        .. code-block:: python

            import os

            from dagster import op, job
            from dagster_msteams import Card, msteams_resource


            @op(required_resource_keys={"msteams"})
            def teams_op(context):
                card = Card()
                card.add_attachment(text_message="Hello There !!")
                context.resources.msteams.post_message(payload=card.payload)


            @job(resource_defs={"msteams": msteams_resource})
            def teams_job():
                teams_op()


            teams_job.execute_in_process(
                {"resources": {"msteams": {"config": {"hook_url": os.getenv("TEAMS_WEBHOOK_URL")}}}}
            )
    """
    return MSTeamsResource.from_resource_context(context).get_client()
