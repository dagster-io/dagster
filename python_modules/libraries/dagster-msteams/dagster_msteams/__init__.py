from dagster_shared.libraries import DagsterLibraryRegistry

from dagster_msteams.card import Card as Card
from dagster_msteams.hooks import (
    teams_on_failure as teams_on_failure,
    teams_on_success as teams_on_success,
)
from dagster_msteams.resources import (
    MSTeamsResource as MSTeamsResource,
    msteams_resource as msteams_resource,
)
from dagster_msteams.sensors import (
    make_teams_on_run_failure_sensor as make_teams_on_run_failure_sensor,
)
from dagster_msteams.version import __version__ as __version__

DagsterLibraryRegistry.register("dagster-msteams", __version__)
