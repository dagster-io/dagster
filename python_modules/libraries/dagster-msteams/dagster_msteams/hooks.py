from typing import Callable, Optional

from dagster._annotations import deprecated_param
from dagster._core.definitions import failure_hook, success_hook
from dagster._core.execution.context.hook import HookContext
from dagster._utils.warnings import normalize_renamed_param

from dagster_msteams.adaptive_card import AdaptiveCard
from dagster_msteams.card import Card
from dagster_msteams.resources import MSTeamsResource
from dagster_msteams.utils import MSTeamsHyperlink, build_message_with_link


def _default_status_message(context: HookContext, status: str) -> str:
    return f"Op {context.op.name} on job {context.job_name} {status}!\nRun ID: {context.run_id}"


def _default_failure_message(context: HookContext) -> str:
    return _default_status_message(context, status="failed")


def _default_success_message(context: HookContext) -> str:
    return _default_status_message(context, status="succeeded")


@deprecated_param(
    param="dagit_base_url",
    breaking_version="2.0",
    additional_warn_text="Use `webserver_base_url` instead.",
)
def teams_on_failure(
    message_fn: Callable[[HookContext], str] = _default_failure_message,
    dagit_base_url: Optional[str] = None,
    webserver_base_url: Optional[str] = None,
):
    """Create a hook on step failure events that will message the given MS Teams webhook URL.

    Args:
        message_fn (Optional(Callable[[HookContext], str])): Function which takes in the
            HookContext outputs the message you want to send.
        dagit_base_url: (Optional[str]): The base url of your webserver instance. Specify this
            to allow messages to include deeplinks to the specific run that triggered
            the hook.
        webserver_base_url: (Optional[str]): The base url of your webserver instance. Specify this
            to allow messages to include deeplinks to the specific run that triggered
            the hook.

    Examples:
        .. code-block:: python

            @teams_on_failure(webserver_base_url="http://localhost:3000")
            @job(...)
            def my_job():
                pass

        .. code-block:: python

            def my_message_fn(context: HookContext) -> str:
                return f"Op {context.op.name} failed!"

            @op
            def a_op(context):
                pass

            @job(...)
            def my_job():
                a_op.with_hooks(hook_defs={teams_on_failure("#foo", my_message_fn)})

    """
    webserver_base_url = normalize_renamed_param(
        webserver_base_url, "webserver_base_url", dagit_base_url, "dagit_base_url"
    )

    @failure_hook(required_resource_keys={"msteams"})
    def _hook(context: HookContext):
        if isinstance(context.resources.msteams, MSTeamsResource):
            client = context.resources.msteams.get_client()
        else:
            client = context.resources.msteams

        message = message_fn(context)
        link = (
            MSTeamsHyperlink("View in Dagster UI", f"{webserver_base_url}/runs/{context.run_id}")
            if webserver_base_url
            else None
        )
        card = Card() if client.is_legacy_webhook() else AdaptiveCard()
        card.add_attachment(build_message_with_link(client.is_legacy_webhook(), message, link))
        client.post_message(card.payload)

    return _hook


@deprecated_param(
    param="dagit_base_url",
    breaking_version="2.0",
    additional_warn_text="Use `webserver_base_url` instead.",
)
def teams_on_success(
    message_fn: Callable[[HookContext], str] = _default_success_message,
    dagit_base_url: Optional[str] = None,
    webserver_base_url: Optional[str] = None,
):
    """Create a hook on step success events that will message the given MS Teams webhook URL.

    Args:
        message_fn (Optional(Callable[[HookContext], str])): Function which takes in the
            HookContext outputs the message you want to send.
        dagit_base_url: (Optional[str]): The base url of your webserver instance. Specify this
            to allow messages to include deeplinks to the specific run that triggered
            the hook.

    Examples:
        .. code-block:: python

            @teams_on_success(webserver_base_url="http://localhost:3000")
            @job(...)
            def my_job():
                pass

        .. code-block:: python

            def my_message_fn(context: HookContext) -> str:
                return f"Op {context.op.name} failed!"

            @op
            def a_op(context):
                pass

            @job(...)
            def my_job():
                a_op.with_hooks(hook_defs={teams_on_success("#foo", my_message_fn)})

    """
    webserver_base_url = normalize_renamed_param(
        webserver_base_url, "webserver_base_url", dagit_base_url, "dagit_base_url"
    )

    @success_hook(required_resource_keys={"msteams"})
    def _hook(context: HookContext):
        if isinstance(context.resources.msteams, MSTeamsResource):
            client = context.resources.msteams.get_client()
        else:
            client = context.resources.msteams

        message = message_fn(context)
        link = (
            MSTeamsHyperlink("View in Dagster UI", f"{webserver_base_url}/runs/{context.run_id}")
            if webserver_base_url
            else None
        )
        card = Card() if client.is_legacy_webhook() else AdaptiveCard()
        card.add_attachment(build_message_with_link(client.is_legacy_webhook(), message, link))
        client.post_message(card.payload)

    return _hook
