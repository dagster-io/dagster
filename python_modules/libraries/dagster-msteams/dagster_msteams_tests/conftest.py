from unittest.mock import patch

import pytest
from dagster_msteams.client import TeamsClient


@pytest.fixture
def json_message():
    test_json_message = {
        "type": "text",
        "attachments": [
            {
                "contentType": "application/vnd.microsoft.card.hero",
                "content": {
                    "title": "Dagster Pipeline Alert",
                    "subtitle": "Hello there !",
                },
            }
        ],
    }
    return test_json_message


@pytest.fixture
def teams_client():
    client = TeamsClient(hook_url="https://some_url_here/")
    return client


@pytest.fixture(scope="function", name="mock_post_method")
def create_mock_post_method():
    with patch("dagster_msteams.client.post") as mock_post:
        mock_response = mock_post.return_value
        mock_response.status_code = 200
        mock_response.json.return_value = {"message": "Success"}
        mock_response.text = "1"
        yield mock_post


LEGACY_WEBHOOK_URL = "https://foo.webhook.office.com/bar/baz"
WEBHOOK_URL = "https://foo.westus.logic.azure.com:443/workflows/8be36cde7f394925af220480f6701bd0"
