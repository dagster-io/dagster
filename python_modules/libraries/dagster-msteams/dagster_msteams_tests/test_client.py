from typing import Any

from dagster_msteams.adaptive_card import AdaptiveCard
from dagster_msteams.card import Card
from dagster_msteams.client import TeamsClient
from dagster_msteams.utils import MSTeamsHyperlink, build_message_with_link

from dagster_msteams_tests.conftest import LEGACY_WEBHOOK_URL, WEBHOOK_URL


def test_client_compatible_with_legacy_webhook(mock_post_method, snapshot: Any):
    teams_client = TeamsClient(hook_url=LEGACY_WEBHOOK_URL)

    card = Card()
    card.add_attachment(
        build_message_with_link(
            is_legacy_webhook=True,
            text="Run failed!",
            link=MSTeamsHyperlink(text="View in Dagit", url="http://localhost:3000"),
        )
    )
    teams_client.post_message(card.payload)

    snapshot.assert_match(mock_post_method.call_args_list)


def test_client_compatible_with_workflow_webhook(mock_post_method, snapshot: Any):
    teams_client = TeamsClient(hook_url=WEBHOOK_URL)

    card = AdaptiveCard()
    card.add_attachment(
        build_message_with_link(
            is_legacy_webhook=False,
            text="Run failed!",
            link=MSTeamsHyperlink(text="View in Dagit", url="http://localhost:3000"),
        )
    )
    teams_client.post_message(card.payload)

    snapshot.assert_match(mock_post_method.call_args_list)
