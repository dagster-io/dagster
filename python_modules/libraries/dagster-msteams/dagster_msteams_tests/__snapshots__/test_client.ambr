# serializer version: 1
# name: test_client_compatible_with_legacy_webhook
  _CallList([
    _Call(
      tuple(
        'https://foo.webhook.office.com/bar/baz',
      ),
      dict({
        'headers': dict({
          'Content-Type': 'application/json',
        }),
        'json': dict({
          'attachments': list([
            dict({
              'content': dict({
                'text': "Run failed! <a href='http://localhost:3000'>View in Dagit</a>",
                'title': 'Dagster Pipeline Alert',
              }),
              'contentType': 'application/vnd.microsoft.card.hero',
            }),
          ]),
          'type': 'message',
        }),
        'proxies': None,
        'timeout': 60,
        'verify': None,
      }),
    ),
  ])
# ---
# name: test_client_compatible_with_workflow_webhook
  _CallList([
    _Call(
      tuple(
        'https://foo.westus.logic.azure.com:443/workflows/8be36cde7f394925af220480f6701bd0',
      ),
      dict({
        'headers': dict({
          'Content-Type': 'application/json',
        }),
        'json': dict({
          'attachments': list([
            dict({
              'content': dict({
                'body': list([
                  dict({
                    'text': 'Run failed! [View in Dagit](http://localhost:3000)',
                    'type': 'TextBlock',
                    'wrap': True,
                  }),
                ]),
                'type': 'AdaptiveCard',
                'version': '1.5',
              }),
              'contentType': 'application/vnd.microsoft.card.adaptive',
              'contentUrl': None,
            }),
          ]),
          'type': 'message',
        }),
        'proxies': None,
        'timeout': 60,
        'verify': None,
      }),
    ),
  ])
# ---
